internal/pipeline/interpolate_matrix.go

```diff
@@ -0,0 +1,57 @@
+package pipeline
+
+import (
+	"fmt"
+	"regexp"
+)
+
+// Match double curly bois containing any whitespace, "matrix", then maybe
+// a dot and a dimension name, ending with any whitespace and closing curlies.
+var matrixTokenRE = regexp.MustCompile(`\{\{\s*matrix(\.[\w-\.]+)?\s*\}\}`)
+
+type stringTransformFunc = func(string) string
+
+// matrixInterpolator creates a reusable string transformer that applies matrix
+// interpolation.
```

Comment:
yeah, good call. i've been thinking vaguely about making `MatrixSelection` its own type (`type MatrixSelection struct { Dimension string, Value any }` or something) - do you think that would provide the necessary clarity?

Reply:
Yeah, I think that helps a lot.

A comment on the type with some examples would centralise the documentation instead of having it in every method with the argument `selection map[string]any`, so I emphatically agree with making it its own type. 

Your proposal of `type MatrixSelection struct { Dimension string, Value any }` illustrates the confusion I was having, though. Is a `selection` a single assignment of a value to a dimension, or is it a collection of them? I think the ordinary usage of "selection" suggests it is a collection, but your type definition suggests it is singular. The argument was `selection map[string]any`, which suggests it's a collection. So there's plenty of ambiguity going around.

I think a term like `MatrixDimensionValue`, while being more boring, does not suffer from this ambiguity. Maybe you should consider
```go
type MatrixDimensionValues map[string]any
```

Selection also suggests that each dimension is assigned a value, but that does not seem to be the case with your `type MatrixSelection struct { Dimension string, Value any }`, and I don't think that's allow in the matrix interpolation notation, e.g., `{{matrix.arch.os}}` is not valid.

Reply:
yeah, when implementing this i realised the same thing you have and arrived at:
```Go
type MatrixSelection []SelectedDimension

type SelectedDimension struct {
  Dimension string
  Value any
}
```

the `any` in SelectedDimension is going to just be a scalar by convention, and enforced at the pipeline level - i agree that we don't want to have nested matrix values

Reply:
(i updated this to be `MatrixPermutation`, which i think is slightly clearer)